HTTP Object submission (#447)
irrdnet/irrd: docs/users/database-changes.rst

IRRd. For each change, a number of integrity, authentication, and reference
checks are performed.
Additionally, notifications may be sent on attempted or successful changes.

.. highlight:: yaml

Submission format
-----------------
There are two ways to submit changes:

* By sending an e-mail with the RPSL objects. This method supports MD5-PW,
  CRYPT-PW and PGPKEY authentication. You will receive a reply by e-mail
  with the result.
* Over HTTPS, through a REST API. This method supports MD5-PW and CRYPT-PW
  authentication. You receive the results in the HTTP response.

All objects submitted are validated for the presence, count and syntax,
though the syntax validation is limited for some attributes.
Values like prefixes are also rewritten into a standard format. If this
results in changes compared to the original submitted text, an info message
is added in the response.

IRRd will attempt to process as many changes as possible, meaning that it's
possible that some changes will fail, and some will succeed. A response will
be sent with the results of the submitted changes, and why any failures
occurred.

Submitting over HTTP
^^^^^^^^^^^^^^^^^^^^

> To submit changes over HTTP, make POST or DELETE requests to ``/v1/submit/``.

Author: ```suggestion
To submit changes over HTTP, make a POST or DELETE request to ``/v1/submit/``.
```

For example, if the IRRd instance is running on ``rr.example.net``, the URL is::

    https://rr.example.net/v1/submit/

The expected request body is a JSON object, with a number of keys:

* ``objects``: a list of objects. Each is a JSON object with
  either a ``object_text`` key containing the full RPSL text,
  or an ``attributes`` key which is a list of objects, each having a
  ``name`` and ``value`` key.
  In other words, you can choose to submit the full object text, or
  the individual attributes separately. The ``value`` key may also be
  a list, which will be translated into RPSL by IRRd.
* ``passwords``: an optional list of passwords to use for authentication.
  Each password will be considered for each object to be changed.
* ``delete_reason``: an optional string with the reason for object deletion.
* ``override``: an optional string containing the override password.

If your request is an HTTP POST, the objects are created or modified, depending
on whether an object already exists with the same primary key and source.
If your request is an HTTP DELETE, the objects are deleted.

> This is a full example of a body that creates/modifies two objects.

Author: ```suggestion
This is an example of a body that creates/modifies two objects.
```

One uses the ``object_text`` option to provide a single string,

> the other specified individual attributes::

Author: ```suggestion
the other specifies individual attributes::
```

    {
        'objects': [
            {
                'object_text': 'person: PERSON1-TEST\n...'
            },
            {
                'attributes': [
                    {
                        'name': 'person',
                        'value': 'PERSON2-TEST'
                    },
                    {
                        'name': 'mnt-by',
                        'value': ['DEMO-TEST', 'DEMO2-TEST']
                    }
                ]
            }
        ],
        'passwords': ['password1', 'password2']
    }

There are two possible responses:

* If there is a syntax error in your JSON object, you will receive
  a ``text/plain`` response with status code 400. The response will

>   tell you what the issue is with your object.

Author: ```suggestion
  tell you what the issue is with your JSON.
```

* If the request was syntactically valid, you always receive a
  ``text/json`` response with status code 200, and the details of
  your change.

> Here is an example of the JSON response::

Author: ```suggestion
Here is an example of a JSON response::
```
